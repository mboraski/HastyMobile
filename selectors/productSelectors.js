import { createSelector } from 'reselect';

const getDeliveryType = state => state.product.deliveryType;

const getProducts = state => state.product.list;

export const getProductsByDeliveryType = createSelector(
    [getDeliveryType, getProducts],
    (deliveryType, products) => {
        const productsByDeliveryType = products[deliveryType] || {};
        return Object.keys(productsByDeliveryType).map(productCode => ({
            ...productsByDeliveryType[productCode],
            productCode,
            deliveryType
        }));
    }
);
