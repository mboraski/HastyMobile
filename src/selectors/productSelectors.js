import { createSelector } from 'reselect';
import filter from 'lodash.filter';
import reduce from 'lodash.reduce';

import { getCartInstantProducts } from './cartSelectors';

export const getHeader = state => state.header;
export const getItemCountUp = state => state.cart.itemCountUp;
export const getItemCountDown = state => state.cart.itemCountDown;

export const getProductsPending = state => state.product.pending;
export const getError = state => state.product.error;
export const getAvailableProducts = state => state.product.instantProducts;
export const getCategory = state => state.product.category.toUpperCase();
export const getProductImages = state => state.product.productImages;

export const getCategories = createSelector(
    [getCartInstantProducts],
    instantProducts => {
        if (instantProducts) {
            const uniqueCategories = reduce(
                instantProducts,
                (accum, product) => {
                    if (accum.indexOf(product.category) === -1) {
                        accum.push(product.category);
                    }
                    return accum;
                },
                []
            );
            console.log(
                'productSelectors; getCategories; uniqueCategories: ',
                uniqueCategories
            );
            return uniqueCategories;
        }
    }
);

export const getProductsByCategory = createSelector(
    [getCategory, getCartInstantProducts],
    (category, instantProducts) => {
        if (instantProducts) {
            return filter(
                instantProducts,
                product => product.category === category.toLowerCase()
            );
        }
    }
);

export const getNumberOfProducts = createSelector(
    [getProductsByCategory],
    products => products.length
);
