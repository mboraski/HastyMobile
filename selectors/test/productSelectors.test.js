import { getProductsByDeliveryType } from '../productSelectors';

describe('productSelector', () => {
    it('getProductsByDeliveryType', () => {
        const state = {
            product: {
                list: {
                    1: {
                        1: {
                            quantity: 1,
                            title: 'Type 1 code 1',
                            thumbnail_image: 'image',
                            price: '$3.49'
                        }
                    },
                    2: {
                        2: {
                            quantity: 2,
                            title: 'Type 2 code 2',
                            thumbnail_image: 'image',
                            price: '$3.49'
                        }
                    }
                },
                deliveryType: '1'
            }
        };
        expect(getProductsByDeliveryType(state)).toMatchSnapshot();
    });
});
