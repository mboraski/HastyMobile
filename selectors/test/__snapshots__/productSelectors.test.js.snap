// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`productSelector getProductsByDeliveryType 1`] = `
Array [
  Object {
    "price": "$3.49",
    "productCode": "1",
    "quantity": 1,
    "thumbnail_image": "image",
    "title": "Type 1 code 1",
  },
]
`;
